STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_registry.c

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#include "_az_ulib_registry_interface.h"
+#include "az_ulib_ipc_api.h"
+#include "az_ulib_ipc_interface.h"
+#include "az_ulib_registry_api.h"
+#include "az_ulib_result.h"
+#include "stm32l475_flash_driver.h"
+#include <azure/core/internal/az_precondition_internal.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+extern uint32_t __REGISTRYINFO;
+extern uint32_t __REGISTRY;
+
+static az_ulib_pal_os_lock registry_lock;
+
+static az_result result_from_hal_status(HAL_StatusTypeDef status)
+{
+  switch (status)
+  {
+    case HAL_OK:
+      return AZ_OK;
+    case HAL_ERROR:
+      return AZ_ERROR_ULIB_SYSTEM;
+    case HAL_BUSY:
+      return AZ_ERROR_ULIB_BUSY;
+    case HAL_TIMEOUT:
+      return AZ_ERROR_ULIB_SYSTEM; // should be AZ_ERROR_ULIB_TIME_OUT
+    default:
+      return AZ_ERROR_ULIB_SYSTEM;
+  }
+}
+
+static az_result set_registry_node_ready_flag(uint8_t* address)
+{
+  AZ_ULIB_TRY
+  {
+    // if ((hal_status = internal_flash_write_doubleword((uint32_t)address, REGISTRY_READY)) !=
+    // HAL_OK)
+    // {
+    //   result = result_from_hal_status(hal_status);
+    //   AZ_ULIB_THROW_IF_AZ_ERROR(result);
+    // }
+    AZ_ULIB_THROW_IF_AZ_ERROR(
+        result_from_hal_status(internal_flash_write_doubleword((uint32_t)address, REGISTRY_READY)));
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result set_registry_node_delete_flag(uint8_t* address)
+{
+  AZ_ULIB_TRY
+  {
+    // if ((hal_status = internal_flash_write_doubleword(
+    //          ((uint32_t)address + AZ_ULIB_REGISTRY_FLAG_SIZE), REGISTRY_DELETED))
+    //     != HAL_OK)
+    // {
+    //   result = result_from_hal_status(hal_status);
+    //   AZ_ULIB_THROW_IF_AZ_ERROR(result);
+    // }
+    AZ_ULIB_THROW_IF_AZ_ERROR(result_from_hal_status(internal_flash_write_doubleword(
+        ((uint32_t)address + AZ_ULIB_REGISTRY_FLAG_SIZE), REGISTRY_DELETED)));
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result store_registry_node(az_ulib_registry_node node, uint8_t** node_ptr)
+{
+  AZ_ULIB_TRY
+  {
+    az_ulib_registry_node* runner;
+    int32_t node_index;
+
+    runner = (az_ulib_registry_node*)(&__REGISTRYINFO);
+
+    /* Look for next available spot */
+    for (node_index = 0; node_index < MAX_AZ_ULIB_REGISTRY_ENTRIES; node_index++)
+    {
+      if (runner->_internal.status.ready_flag == REGISTRY_FREE)
```

Comment:
>if (runner->_internal.status.ready_flag == REGISTRY_FREE) [](http://example.com/codeflow?start=6&length=57)

A node may be compromised by a power down of a crash even with the ready_flasg set to REGISTRY_FREE, so you need to test the entire node and only use it if all bits are set to `1`.